Test/lambdas/StateMonad.dfy

```diff
@@ -0,0 +1,231 @@
+
+  // (m >>= f) >>= g = m >>= (x => f(x) >>= g)
+  lemma Associativity<A,B,C>(m : M<A>, f:A -> M<B>, g: B -> M<C>)
+    ensures Bind(Bind(m,f),g) ==
+            Bind(m,x => Bind(f(x),g))
+}
+
+module State refines Monad {
+//   type S // doesn't work because S is used left of an arrow
+  datatype M<A> = State(runState: int -> (A, int))
+
+  function method Return<A>(x: A): M<A>
+  { State(s => (x, s)) }
+
+  function method Bind<A,B>(m: M<A>, f:A -> M<B>): M<B>
+  {
+    match m case State(h) =>
+      State(s =>
+            match h(s) case (a, newState) =>
+              match f(a) case State(g) => g(newState))
+  }
+
+  lemma FunEq<X, Y>(f: X -> Y, g: X -> Y)
+    requires forall x :: f(x) == g(x) 
+    ensures f == g 
+  { assume false; }
+
+  lemma LeftIdentity<A,B>(x : A, f : A -> M<B>)
+    ensures Bind(Return(x), f) == f(x) // return x >>= f  ==  f x
+  {
+      var State(h) := State(s => (x, s));
+      var State(g2) := f(x);
+      calc {
+          Bind(Return(x), f);
+          ==
+          Bind(State(s => (x, s)), f);
+          ==
+          State(s =>
+                match ((s => (x, s))(s))
+                case (a, newState) =>
+                  match f(a)
+                  case State(g) => g(newState));
+          == { FunEq(s => match ((s => (x, s))(s))
+                case (a, newState) =>
+                  match f(a)
+                  case State(g) => g(newState), 
+                s => match (x, s)
+                case (a, newState) =>
+                  match f(a)
+                  case State(g) => g(newState)); }
+          State(s =>
+                match (x, s)
+                case (a, newState) =>
+                  match f(a)
+                  case State(g) => g(newState));
+          == {
+               FunEq((s =>
+                match (x, s)
+                case (a, newState) =>
+                  match f(a)
+                  case State(g) => g(newState)), 
+                     (s => match f(x) case State(g) => g(s)));
+             }
+          State(s => match f(x) case State(g) => g(s));
+          == { FunEq(s => match f(x) case State(g) => g(s), g2); }
+          State(g2);
+          ==
+          f(x);
+    }
+  }
+
+  lemma RightIdentity<A>(m : M<A>)
+    {
+        match m case State(h) =>
+        calc {
+          Bind(m, x => Return(x));
+          ==
+          State(s =>
+                    match h(s)
+                    case (a, newState) =>
+                      match (x => Return(x))(a)
+                        case State(g) => g(newState));
+          == {
+            FunEq((s => match h(s) case (a, newState) =>
+                          match (x => Return(x))(a) case State(g) => g(newState)), 
+                  (s => h(s)));
+          }
+          State(s => h(s));
+          == { FunEq(s => h(s), h); }
+          State(h);
+          ==
+          m;
+        }
+    }
+
+  lemma Associativity<A,B,C>(m : M<A>, f:A -> M<B>, g: B -> M<C>)
```

Comment:
Not a criticism, but curious: is there a path to making this kind of proof less verbose?  I'm not used to seeing so many manual expansions in calc sequences.  Do you think this is a fundamental limitation, or do we just need more axioms or triggers or something in the way Dafny automates other features.

Reply:
I think that part of what makes the proof so verbose is that we are reasoning about nested lambdas each of which we have to write out in full because the verifier doesn't reduce anything inside of lambda bodies on its own. So not only do we have to keep writing everything fully, but we have to also copy-paste the lambdas into the functional-extensionality-axiom application `FunEq`.
According to @RustanLeino even when I add `FunEq` to Dafny by default (so that we can write `f == g` rather than `FunEq(f, g)`) we will still have to explicitly give hints to the verifier that it should invoke functional extensionality, which will not make the proofs much less verbose. Although I do hope that it'll be able to apply the axiom on its own, at least sometimes, but I am still in the process of adding extensionality by default so I don't know yet.